File: backend/models/schedule.py
"""Schedule output data models.

These models define the shape of a generated schedule plan —
the scored, analyzed output that the frontend receives.
"""

from typing import Optional

from pydantic import BaseModel, Field

from models.course import Course


class ScoreBreakdown(BaseModel):
    """Schedule-focused scoring dimensions + legacy stubs.

    All scores in [1.0, 10.0]. Higher is always better.
    """

    # Current schedule-optimizer dimensions
    free_days: float = Field(default=5.0, description="Free whole days score (1.0-10.0)")
    compactness: float = Field(default=5.0, description="Schedule compactness (1.0-10.0)")
    no_morning: float = Field(default=5.0, description="No-early-morning score (10.0 = perfect)")
    no_night: float = Field(default=5.0, description="No-night-class score (10.0 = perfect)")
    distribution: float = Field(default=5.0, description="Daily balance score (10.0 = even spread)")

    # Legacy stubs (reserved, always 5.0)
    gpa_score: float = Field(default=5.0, description="[Legacy] GPA score stub")
    compact_score: float = Field(default=5.0, description="[Legacy] Compactness stub")
    stress_score: float = Field(default=5.0, description="[Legacy] Stress score stub")
    free_score: float = Field(default=5.0, description="[Legacy] Free half-day score stub")
    morning_penalty: float = Field(default=5.0, description="[Legacy] Morning penalty stub")
    friday_penalty: float = Field(default=5.0, description="[Legacy] Friday penalty stub")
    monday_penalty: float = Field(default=5.0, description="[Legacy] Monday penalty stub")
    afternoon_penalty: float = Field(default=5.0, description="[Legacy] Afternoon penalty stub")


class PlanAnalysis(BaseModel):
    """Aggregated metadata and analysis for one schedule plan."""

    total_credits: int = Field(..., description="Sum of all course credits")
    school_days: int = Field(..., description="Number of days with at least one class")
    earliest_period: int = Field(
        ..., description="Earliest period number across the week"
    )
    score_breakdown: ScoreBreakdown
    reasons: list[str] = Field(
        default_factory=list, description="Natural language recommendation reasons"
    )


class Plan(BaseModel):
    """One complete ranked schedule recommendation."""

    id: str = Field(..., description="Plan identifier: plan_A, plan_B, plan_C")
    rank: int = Field(..., ge=1, le=3, description="Rank: 1 (best), 2, or 3")
    score: float = Field(..., description="Weighted total score (0-10)")
    label: str = Field(
        ..., description="Display label: 推荐方案 A / 备选方案 B / 备选方案 C"
    )
    courses: list[Course] = Field(..., description="Selected course sections")
    analysis: PlanAnalysis
    matched_strategies: list[str] = Field(
        default_factory=list,
        description="Strategy IDs that apply to this plan",
    )
    llm_review: Optional[dict] = Field(
        default=None,
        description="LLM evaluation (only when llm_evaluate=True)"
    )

File: frontend/src/types/schedule.ts
// Mirrors backend models/course.py ScheduleSlot
export interface ScheduleSlot {
  day: number;   // 1=Monday .. 5=Friday
  start: number; // 1-14
  end: number;   // 1-14, inclusive
}

// Mirrors backend models/course.py Teacher
export interface Teacher {
  name: string;
  rating: number; // 1.0-5.0
}

// Mirrors backend models/course.py Location
export interface Location {
  building: string;
  floor: number;
  room: string;
}

// Mirrors backend models/course.py Course
export interface Course {
  section_id: string;
  course_code: string;
  course_name: string;
  credit_transfer_group: string;
  credits: number;
  teacher: Teacher | null;
  schedule: ScheduleSlot[];
  location: Location | null;
  course_type: "major" | "easy";
  category: "regular" | "pe";
  required: boolean;
  delivery_mode: "线下传统" | "线上网课" | "线上线下混合";
  semester: string;
}

// Mirrors backend models/schedule.py ScoreBreakdown
export interface ScoreBreakdown {
// Schedule-focused metrics (current)
  free_days: number;
  compactness: number;
  no_morning: number;
  no_night: number;
  distribution: number;
  // Legacy fields (reserved for future GPA/teacher modules)
  gpa_score?: number;
  compact_score?: number;
  stress_score?: number;
  free_score?: number;
  morning_penalty?: number;
  friday_penalty?: number;
  monday_penalty?: number;
  afternoon_penalty?: number;
}

// Mirrors backend models/schedule.py PlanAnalysis
export interface PlanAnalysis {
  total_credits: number;
  school_days: number;
  earliest_period: number;
  score_breakdown: ScoreBreakdown;
  reasons: string[];
}

// Mirrors backend models/schedule.py Plan
export interface Plan {
  id: string;
  rank: number;
  score: number;
  label: string;
  courses: Course[];
  analysis: PlanAnalysis;
  matched_strategies: string[];
  llm_review?: {
    overall: string;
    score: number;
    pros: string[];
    cons: string[];
    suggestions: string[];
    best_for: string;
  } | null; // strategy IDs, not full objects
}

// Mirrors backend models/response.py GenerateResponse
export interface GenerateResponse {
  scenario: string;
  plans: Plan[];
}

// Mirrors backend models/response.py ErrorResponse
export interface ErrorResponse {
  error: string;
  detail: string;
  message: string;
}

// Mirrors backend routes/import_router.py ImportStats + ImportResponse
export interface ImportStats {
  course_count: number;
  processing_time_ms: number;
  method: string;
}

export interface ImportResult {
  success: boolean;
  method: string;
  courses: Course[];
  stats: ImportStats;
}

export interface ImportError {
  success: false;
  error: string;
  message: string;
}
